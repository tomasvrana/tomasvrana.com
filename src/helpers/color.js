import TinyColor from '@ctrl/tinycolor'
import logger from './logger'

const makeTransparent = (color, alpha) => {
  try {
    const c = new TinyColor(color)
    c.setAlpha(alpha)
    return c.toRgbString()
  } catch (error) {
    logger.error({
      description: 'error making color transparent',
      args: [ color, alpha ],
      error
    })
    return '#000000'
  }
}

export {
  makeTransparent
}
