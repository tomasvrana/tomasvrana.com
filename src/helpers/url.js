import { triggerEvent } from './dom'
import config from '../config'

const window = global.window || { location: { pathname: '/', href: '', origin: '', hash: '', hostname: '' }, history: {} }

export const getURIPath = () => window.location.pathname
export const getURIPathWithoutLang = lang => getURIPath().replace(new RegExp(`^\\/${lang}`), '')
export const languageSetInURL = (url = null) => !!config.availableLanguages[(url || getURIPath()).split('/').splice(1)[0]]
export const getHref = (path = null) => path ? `${window.location.origin}${path}` : window.location.href
export const setHref = href => (window.location.href = href)
export const isRelative = url => !url.match(/^(\/\/|https?:\/\/|mailto:)/)
export const isPath = url => !url.match(/^(\/|\/[^\\/].*)$/)
export const getHash = () => window.location.hash
export const getHostname = () => window.location.hostname

const getHashParams = () => {
  const currentHash = getHash().replace(/^#/, '')
  const params = currentHash.split('&').map(p => ({ [p.split('=')[0]]: p.split('=')[1] }))
  let paramsObj = {}
  params.forEach(p => (paramsObj = { ...paramsObj, ...p }))
  return paramsObj
}

export const setHashParam = (key, value) => {
  const params = getHashParams()
  if (!value) {
    delete params[key]
  } else {
    params[key] = value
  }

  const paramStrings = Object.keys(params).map(k => params[k] && k ? `${k}=${params[k]}` : '').filter(v => !!v)
  if (paramStrings.length > 0) {
    if (window.history.pushState) {
      window.history.pushState(null, null, `#${paramStrings.join('=')}`)
      triggerEvent({ type: 'hashchange' })
    } else {
      window.location.hash = `#${paramStrings.join('=')}`
    }
  } else {
    if (window.history.pushState) {
      window.history.pushState(null, null, '#')
      triggerEvent({ type: 'hashchange' })
    } else {
      window.location.hash = ''
    }
  }
}

export const getHashParam = key => {
  return getHashParams()[key] || ''
}
