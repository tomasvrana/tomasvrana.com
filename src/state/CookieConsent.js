import React, { createContext, useState } from 'react'
import PropTypes from 'prop-types'
import { getLocalStorageValue, setLocalStorageValue } from '../helpers/localStorage'
import { allowTracking } from '../helpers/analytics'

const CookieConsentContext = createContext()
const Consumer = CookieConsentContext.Consumer

const Provider = ({ children }) => {
  const [ allowCookies, setAllowCookies ] = useState(getLocalStorageValue('allowCookies') === 'true')

  const updateAllowCookies = allow => {
    if (allow) {
      allowTracking()
      setLocalStorageValue('allowCookies', 'true')
    }

    setAllowCookies(allow)
  }

  return (
    <CookieConsentContext.Provider value={{ allowCookies, setAllowCookies: updateAllowCookies }}>
      {children}
    </CookieConsentContext.Provider>
  )
}

Provider.propTypes = {
  children: PropTypes.any
}

export {
  Provider,
  Consumer
}
