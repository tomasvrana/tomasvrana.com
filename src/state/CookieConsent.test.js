import React from 'react'
import { mount } from 'enzyme'
import { Provider, Consumer } from './CookieConsent'
import * as localStorage from '../helpers/localStorage'
import * as analytics from '../helpers/analytics'

const getProvider = props => {
  return mount(
    <Provider
      children={<p>Children</p>}
      {...props}
    />
  )
}

describe('CookieConsent', () => {
  beforeEach(() => {
    localStorage.getLocalStorageValue = jest.fn(() => '')
    localStorage.setLocalStorageValue = jest.fn()
    analytics.allowTracking = jest.fn()
  })

  test('the Provider renders correctly', () => {
    expect(getProvider()).toMatchSnapshot()
  })

  describe('allowing cookies', () => {
    let wrapper
    beforeEach(() => {
      wrapper = getProvider({
        children: (
          <Consumer>
            {({ allowCookies, setAllowCookies }) => (
              <div>
                <button onClick={() => setAllowCookies(true)} />
                {allowCookies &&
                  <p>Cookies allowed</p>
                }
              </div>
            )}
          </Consumer>
        )
      })

      wrapper.find('button').first().simulate('click')
    })

    test('it allows tracking', () => {
      expect(analytics.allowTracking).toBeCalled()
    })

    test('it sets the allowCookies local storage value to true', () => {
      expect(localStorage.setLocalStorageValue).toBeCalledWith('allowCookies', 'true')
    })

    test('it renders correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('disallowing cookies', () => {
    let wrapper
    beforeEach(() => {
      wrapper = getProvider({
        children: (
          <Consumer>
            {({ allowCookies, setAllowCookies }) => (
              <div>
                <button onClick={() => setAllowCookies(false)} />
                {allowCookies &&
                  <p>Cookies allowed</p>
                }
              </div>
            )}
          </Consumer>
        )
      })

      wrapper.find('button').first().simulate('click')
    })

    test('it does not allow tracking', () => {
      expect(analytics.allowTracking).not.toBeCalled()
    })

    test('it does not update the local storage value of allowCookies', () => {
      expect(localStorage.setLocalStorageValue).not.toBeCalled()
    })

    test('it renders correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when the allowCookies value in local storage is true', () => {
    let wrapper
    beforeEach(() => {
      localStorage.getLocalStorageValue.mockImplementation((key) => key === 'allowCookies' && 'true')
      wrapper = getProvider({
        children: (
          <Consumer>
            {({ allowCookies }) => (
              <div>
                {allowCookies &&
                  <p>Cookies allowed via local storage</p>
                }
              </div>
            )}
          </Consumer>
        )
      })
    })

    test('it renders correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })
})
