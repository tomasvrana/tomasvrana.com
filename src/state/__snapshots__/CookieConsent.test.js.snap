// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CookieConsent allowing cookies it renders correctly 1`] = `
<Provider>
  <div>
    <button
      onClick={[Function]}
    />
    <p>
      Cookies allowed
    </p>
  </div>
</Provider>
`;

exports[`CookieConsent disallowing cookies it renders correctly 1`] = `
<Provider>
  <div>
    <button
      onClick={[Function]}
    />
  </div>
</Provider>
`;

exports[`CookieConsent the Provider renders correctly 1`] = `
<Provider>
  <p>
    Children
  </p>
</Provider>
`;

exports[`CookieConsent when the allowCookies value in local storage is true it renders correctly 1`] = `
<Provider>
  <div>
    <p>
      Cookies allowed via local storage
    </p>
  </div>
</Provider>
`;
